core/src/main/java/io/questdb/cairo/wal/seq/TableSequencerImpl.java

```diff
@@ -0,0 +1,280 @@
+
+import io.questdb.cairo.*;
+import io.questdb.cairo.wal.WalUtils;
+import io.questdb.griffin.SqlException;
+import io.questdb.log.Log;
+import io.questdb.log.LogFactory;
+import io.questdb.std.FilesFacade;
+import io.questdb.std.Misc;
+import io.questdb.std.SimpleReadWriteLock;
+import io.questdb.std.str.Path;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.concurrent.locks.ReadWriteLock;
+
+import static io.questdb.cairo.wal.WalUtils.WAL_INDEX_FILE_NAME;
+
+public class TableSequencerImpl implements TableSequencer {
+    private static final Log LOG = LogFactory.getLog(TableSequencerImpl.class);
+    private final ReadWriteLock schemaLock = new SimpleReadWriteLock();
+    private final CairoEngine engine;
+    private final String tableName;
+    private final int rootLen;
+    private final SequencerMetadata metadata;
+    private final TableTransactionLog tableTransactionLog;
+    private final IDGenerator walIdGenerator;
+    private final Path path;
+    private final BinaryAlterFormatter alterCommandWalFormatter = new BinaryAlterFormatter();
+    private final SequencerMetadataUpdater sequencerMetadataUpdater;
+    private final FilesFacade ff;
+    private final int mkDirMode;
+    private volatile boolean open = false;
+    private boolean distressed;
+
+    TableSequencerImpl(CairoEngine engine, String tableName) {
+        this.engine = engine;
+        this.tableName = tableName;
+
+        final CairoConfiguration configuration = engine.getConfiguration();
+        final FilesFacade ff = configuration.getFilesFacade();
+        try {
+            path = new Path();
+            path.of(configuration.getRoot()).concat(tableName).concat(WalUtils.SEQ_DIR);
+            rootLen = path.length();
+            this.ff = ff;
+            this.mkDirMode = configuration.getMkDirMode();
+
+            metadata = new SequencerMetadata(ff);
+            sequencerMetadataUpdater = new SequencerMetadataUpdater(metadata, tableName);
+            walIdGenerator = new IDGenerator(configuration, WAL_INDEX_FILE_NAME);
+            tableTransactionLog = new TableTransactionLog(ff);
+        } catch (Throwable th) {
+            LOG.critical().$("could not create sequencer [name=").$(tableName)
+                    .$(", error=").$(th.getMessage())
+                    .I$();
+            doClose();
+            throw th;
+        }
+    }
+
+    @Override
+    public void close() {
+        schemaLock.writeLock().lock();
+        try {
+            doClose();
+        } finally {
+            schemaLock.writeLock().unlock();
+        }
+    }
+
+    @Override
+    public void copyMetadataTo(@NotNull SequencerMetadata copyTo) {
+        schemaLock.readLock().lock();
+        try {
+            copyTo.copyFrom(metadata);
+        } finally {
+            schemaLock.readLock().unlock();
+        }
+    }
+
+    @Override
+    public TableMetadataChangeLog getMetadataChangeLogCursor(long structureVersionLo) {
+        checkDistressed();
```

Comment:
This check is racy.

Reply:
What's meant here is that `TableSequencerImpl#distressed` field should have a `volatile` modifier.

Reply:
Broken multi-threaded code isn't fine. Adding `volatile` modifier doesn't cost us much (in fact, it might be hard to even notice the overhead).

Reply:
`volatile` does not solve the race. Table can become distressed any moment after checking the volatile variable.

Reply:
Absence of `volatile` makes this code much more broken from the synchronization perspective. Is there a reason to avoid having the modifier on the `distressed` field?

Reply:
It does not solve the race and make code confusing (as if solves the race).
Anyway to illuminate the chances of using distressed TableSequencer I moved the check for distressed flag to `TableSequencerAPI` and re-create attempt there.